File: include/c#/10/UtilStructs.cs
namespace Hardstuck.GuildWars2.BuildCodes.V2.Util;

public struct SpecializationChoices {
	public Specialization? Choice1;
	public Specialization? Choice2;
	public Specialization? Choice3;

	public Specialization? this[int index] {
		get => (index) switch {
			0 => this.Choice1,
			1 => this.Choice2,
			2 => this.Choice3,
			_ => throw new ArgumentOutOfRangeException(nameof(index)),
		};
		set {
			switch(index) {
				case 0: this.Choice1 = value; break;
				case 1: this.Choice2 = value; break;
				case 2: this.Choice3 = value; break;
				default: throw new ArgumentOutOfRangeException(nameof(index));
			};
		}
	}
}

public struct TraitLineChoices {
	public TraitLineChoice Adept;
	public TraitLineChoice Master;
	public TraitLineChoice Grandmaster;

	public TraitLineChoice this[int index] {
		get => (index) switch {
			0 => this.Adept,
			1 => this.Master,
			2 => this.Grandmaster,
			_ => throw new ArgumentOutOfRangeException(nameof(index)),
		};
		set {
				switch(index) {
				case 0: this.Adept       = value; break;
				case 1: this.Master      = value; break;
				case 2: this.Grandmaster = value; break;
				default: throw new ArgumentOutOfRangeException(nameof(index));
			};
		}
	}
}

public struct AllWeapons {
	public WeaponSet Set1;
	public WeaponSet Set2;
}

public struct AllSkills {
	public SkillId? Heal;
	public SkillId? Utility1;
	public SkillId? Utility2;
	public SkillId? Utility3;
	public SkillId? Elite;

	public SkillId? this[int index] {
		get => (index) switch {
			0 => this.Heal,
			1 => this.Utility1,
			2 => this.Utility2,
			3 => this.Utility3,
			4 => this.Elite,
			_ => throw new ArgumentOutOfRangeException(nameof(index)),
		};
		set {
			switch(index) {
				case 0: this.Heal     = value; break;
				case 1: this.Utility1 = value; break;
				case 2: this.Utility2 = value; break;
				case 3: this.Utility3 = value; break;
				case 4: this.Elite    = value; break;
				default: throw new ArgumentOutOfRangeException(nameof(index));
			};
		}
	}
}

public struct AllEquipmentStats {
	public StatId  Helmet;
	public StatId  Shoulders;
	public StatId  Chest;
	public StatId  Gloves;
	public StatId  Leggings;
	public StatId  Boots;
	public StatId  BackItem;
	public StatId  Accessory1;
	public StatId  Accessory2;
	public StatId  Ring1;
	public StatId  Ring2;
	public StatId? WeaponSet1MainHand;
	public StatId? WeaponSet1OffHand;
	public StatId? WeaponSet2MainHand;
	public StatId? WeaponSet2OffHand;
	public StatId  Amulet;

	public StatId? this[int index] {
		get => (index) switch {
			 0 => this.Helmet,
			 1 => this.Shoulders,
			 2 => this.Chest,
			 3 => this.Gloves,
			 4 => this.Leggings,
			 5 => this.Boots,
			 6 => this.BackItem,
			 7 => this.Accessory1,
			 8 => this.Accessory2,
			 9 => this.Ring1,
			10 => this.Ring2,
			11 => this.WeaponSet1MainHand,
			12 => this.WeaponSet1OffHand ,
			13 => this.WeaponSet2MainHand,
			14 => this.WeaponSet2OffHand ,
			15 => this.Amulet,
			_ => throw new ArgumentOutOfRangeException(nameof(index)),
		};
		set {
			switch(index) {
				case  0: this.Helmet             = value ?? 0; break;
				case  1: this.Shoulders          = value ?? 0; break;
				case  2: this.Chest              = value ?? 0; break;
				case  3: this.Gloves             = value ?? 0; break;
				case  4: this.Leggings           = value ?? 0; break;
				case  5: this.Boots              = value ?? 0; break;
				case  6: this.BackItem           = value ?? 0; break;
				case  7: this.Accessory1         = value ?? 0; break;
				case  8: this.Accessory2         = value ?? 0; break;
				case  9: this.Ring1              = value ?? 0; break;
				case 10: this.Ring2              = value ?? 0; break;
				case 11: this.WeaponSet1MainHand = value; break;
				case 12: this.WeaponSet1OffHand  = value; break;
				case 13: this.WeaponSet2MainHand = value; break;
				case 14: this.WeaponSet2OffHand  = value; break;
				case 15: this.Amulet             = value ?? 0; break;
				default: throw new ArgumentOutOfRangeException(nameof(index));
			};
		}
	}
}

public struct AllEquipmentInfusions {
	public int? Helmet;
	public int? Shoulders;
	public int? Chest;
	public int? Gloves;
	public int? Leggings;
	public int? Boots;
	public int? BackItem_1;
	public int? BackItem_2;
	public int? Accessory1;
	public int? Accessory2;
	public int? Ring1_1;
	public int? Ring1_2;
	public int? Ring1_3;
	public int? Ring2_1;
	public int? Ring2_2;
	public int? Ring2_3;
	public int? WeaponSet1_1;
	public int? WeaponSet1_2;
	public int? WeaponSet2_1;
	public int? WeaponSet2_2;
	public int? Amulet;

	public int? this[int index]
	{
		get => (index) switch {
			 0 => this.Helmet,
			 1 => this.Shoulders,
			 2 => this.Chest,
			 3 => this.Gloves,
			 4 => this.Leggings,
			 5 => this.Boots,
			 6 => this.BackItem_1,
			 7 => this.BackItem_2,
			 8 => this.Accessory1,
			 9 => this.Accessory2,
			10 => this.Ring1_1,
			11 => this.Ring1_2,
			12 => this.Ring1_3,
			13 => this.Ring2_1,
			14 => this.Ring2_2,
			15 => this.Ring2_3,
			16 => this.WeaponSet1_1,
			17 => this.WeaponSet1_2,
			18 => this.WeaponSet2_1,
			19 => this.WeaponSet2_2,
			20 => this.Amulet,
			_ => throw new ArgumentOutOfRangeException(nameof(index)),
		};

		set {
			switch(index)
			{
				case  0: this.Helmet       = value; break;
				case  1: this.Shoulders    = value; break;
				case  2: this.Chest        = value; break;
				case  3: this.Gloves       = value; break;
				case  4: this.Leggings     = value; break;
				case  5: this.Boots        = value; break;
				case  6: this.BackItem_1   = value; break;
				case  7: this.BackItem_2   = value; break;
				case  8: this.Accessory1   = value; break;
				case  9: this.Accessory2   = value; break;
				case 10: this.Ring1_1      = value; break;
				case 11: this.Ring1_2      = value; break;
				case 12: this.Ring1_3      = value; break;
				case 13: this.Ring2_1      = value; break;
				case 14: this.Ring2_2      = value; break;
				case 15: this.Ring2_3      = value; break;
				case 16: this.WeaponSet1_1 = value; break;
				case 17: this.WeaponSet1_2 = value; break;
				case 18: this.WeaponSet2_1 = value; break;
				case 19: this.WeaponSet2_2 = value; break;
				case 20: this.Amulet       = value; break;
				default: throw new ArgumentOutOfRangeException(nameof(index));
			};
		}
	}
}

File: include/ts/Util/UtilStructs.ts
import { Specialization, TraitLineChoice } from "../Structures";

import ItemId from "../Database/ItemIds";
import StatId from "../Database/StatIds";
import SkillId from "../Database/SkillIds";

export class SpecializationChoices {
	public Choice1 : Specialization = new Specialization();
	public Choice2 : Specialization = new Specialization();
	public Choice3 : Specialization = new Specialization();

	public [0] = this.Choice1;
	public [1] = this.Choice2;
	public [2] = this.Choice3;
}

export class TraitLineChoices {
	public Adept       : TraitLineChoice = TraitLineChoice.NONE;
	public Master      : TraitLineChoice = TraitLineChoice.NONE;
	public Grandmaster : TraitLineChoice = TraitLineChoice.NONE;

	public [0] = this.Adept;
	public [1] = this.Master;
	public [2] = this.Grandmaster;
}

export class AllSkills {
	public Heal     : SkillId = SkillId._UNDEFINED;
	public Utility1 : SkillId = SkillId._UNDEFINED;
	public Utility2 : SkillId = SkillId._UNDEFINED;
	public Utility3 : SkillId = SkillId._UNDEFINED;
	public Elite    : SkillId = SkillId._UNDEFINED;

	public [0] = this.Heal;
	public [1] = this.Utility1;
	public [2] = this.Utility2;
	public [3] = this.Utility3;
	public [4] = this.Elite;
}

export class AllEquipmentStats {
	public Helmet             : StatId = StatId._UNDEFINED;
	public Shoulders          : StatId = StatId._UNDEFINED;
	public Chest              : StatId = StatId._UNDEFINED;
	public Gloves             : StatId = StatId._UNDEFINED;
	public Leggings           : StatId = StatId._UNDEFINED;
	public Boots              : StatId = StatId._UNDEFINED;
	public BackItem           : StatId = StatId._UNDEFINED;
	public Accessory1         : StatId = StatId._UNDEFINED;
	public Accessory2         : StatId = StatId._UNDEFINED;
	public Ring1              : StatId = StatId._UNDEFINED;
	public Ring2              : StatId = StatId._UNDEFINED;
	/** @remarks Is StatId._UNDEFINED if the weapon is not set. */
	public WeaponSet1MainHand : StatId = StatId._UNDEFINED;
	/** @remarks Is StatId._UNDEFINED if the weapon is not set. */
	public WeaponSet1OffHand  : StatId = StatId._UNDEFINED;
	/** @remarks Is StatId._UNDEFINED if the weapon is not set. */
	public WeaponSet2MainHand : StatId = StatId._UNDEFINED;
	/** @remarks Is StatId._UNDEFINED if the weapon is not set. */
	public WeaponSet2OffHand  : StatId = StatId._UNDEFINED;
	public Amulet             : StatId = StatId._UNDEFINED;

	public [ 0] = this.Helmet;
	public [ 1] = this.Shoulders;
	public [ 2] = this.Chest;
	public [ 3] = this.Gloves;
	public [ 4] = this.Leggings;
	public [ 5] = this.Boots;
	public [ 6] = this.BackItem;
	public [ 7] = this.Accessory1;
	public [ 8] = this.Accessory2;
	public [ 9] = this.Ring1;
	public [10] = this.Ring2;
	public [11] = this.WeaponSet1MainHand;
	public [12] = this.WeaponSet1OffHand ;
	public [13] = this.WeaponSet2MainHand;
	public [14] = this.WeaponSet2OffHand ;
	public [15] = this.Amulet;
}

export class AllEquipmentInfusions {
	public Helmet       : ItemId = ItemId._UNDEFINED;
	public Shoulders    : ItemId = ItemId._UNDEFINED;
	public Chest        : ItemId = ItemId._UNDEFINED;
	public Gloves       : ItemId = ItemId._UNDEFINED;
	public Leggings     : ItemId = ItemId._UNDEFINED;
	public Boots        : ItemId = ItemId._UNDEFINED;
	public BackItem_1   : ItemId = ItemId._UNDEFINED;
	public BackItem_2   : ItemId = ItemId._UNDEFINED;
	public Accessory1   : ItemId = ItemId._UNDEFINED;
	public Accessory2   : ItemId = ItemId._UNDEFINED;
	public Ring1_1      : ItemId = ItemId._UNDEFINED;
	public Ring1_2      : ItemId = ItemId._UNDEFINED;
	public Ring1_3      : ItemId = ItemId._UNDEFINED;
	public Ring2_1      : ItemId = ItemId._UNDEFINED;
	public Ring2_2      : ItemId = ItemId._UNDEFINED;
	public Ring2_3      : ItemId = ItemId._UNDEFINED;
	public WeaponSet1_1 : ItemId = ItemId._UNDEFINED;
	public WeaponSet1_2 : ItemId = ItemId._UNDEFINED;
	public WeaponSet2_1 : ItemId = ItemId._UNDEFINED;
	public WeaponSet2_2 : ItemId = ItemId._UNDEFINED;
	public Amulet       : ItemId = ItemId._UNDEFINED;

	public [ 0] = this.Helmet;
	public [ 1] = this.Shoulders;
	public [ 2] = this.Chest;
	public [ 3] = this.Gloves;
	public [ 4] = this.Leggings;
	public [ 5] = this.Boots;
	public [ 6] = this.BackItem_1;
	public [ 7] = this.BackItem_2;
	public [ 8] = this.Accessory1;
	public [ 9] = this.Accessory2;
	public [10] = this.Ring1_1;
	public [11] = this.Ring1_2;
	public [12] = this.Ring1_3;
	public [13] = this.Ring2_1;
	public [14] = this.Ring2_2;
	public [15] = this.Ring2_3;
	public [16] = this.WeaponSet1_1;
	public [17] = this.WeaponSet1_2;
	public [18] = this.WeaponSet2_1;
	public [19] = this.WeaponSet2_2;
	public [20] = this.Amulet;

	//NOTE(Rennorb): It isn't really optimal to use this performance wise, but its very convenient.
	public HasAny() : boolean
	{
		for(let i = 0; i < Static.ALL_INFUSION_COUNT; i++)
			if(this[i] !== ItemId._UNDEFINED)
				return true;
		return false;
	}
}
